Record a response status on the AsyncResult object.

alexjeffburke opened the pull request:
## Description
Some services that communicate via celery jobs may not send back a
result. In order to allow distinguishing successful responses from
failures, the message status must be available.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
The message status can only be inferred (in some cases ambiguously)
from the resolution value of the `.get()` method.

* **What is the new behavior (if this is a feature change)?**
The message status is recorded in the AsyncResult object.

* **Other information**:
As part of this, a job timeout occurring is also exposed via a state
and a test is added for the timeout kicking in.

actumn commented on src/app/result.ts:
```diff
@@ -1,8 +1,11 @@
 import { CeleryBackend } from "../backends";
 
+type AsyncResultStatus = "PENDING" | "SUCCESS" | "FAILURE" | "TIMEOUT";
```
Can you remove "TIMEOUT" in status? I guess there's no need to represent timeout as status.

actumn commented on src/app/result.ts:
```diff
@@ -33,6 +37,7 @@ export class AsyncResult {
       if (timeout) {
         timeoutId = setTimeout(() => {
           clearInterval(intervalId);
+          this.status = "TIMEOUT";
```
How about using `reject(error)` to represent "timeout" rather than just `resolve(null)`?
```suggestion
          reject(new Error("timeout"));
```
Then we can handle timeout without status.
```javascript
result.get(500).then(data => {
  console.log(data);
}).catch(e => {
  console.log(e.message);
});
```

actumn reviewed (changes requested):
Thank you for a nice proposal supporting the result status and the PR is looking good. But I think it could be better so I left a few messages.

alexjeffburke commented:
Thanks for the review. I'm going to be off work for a bit, but happy to rework this when I return.

Changing timeout to be a rejection makes sense :) I only made it `resolve()` with a `status` in that case so that it behaved in the same way as celery returning a FAILURE message. My reasoning was that for consistency in the caller it felt better to reject in that case too, and then you know that if you ever end up in `catch()` that something went wrong.. I guess that could allow completely removing the states. What would you prefer?

actumn commented:
Sorry for the delayed reply. 

As I understand, you hope when a timeout occurs, it works like `FAILURE` in celery and marks status as `TIMEOUT`. But I guess it could be confusing to users because there's no timeout state in celery. 
When I designed `AsyncResult`, I just wanted it to get the result from the result backend if it exists or throws an error as celery works.

And I'm not sure how something goes wrong and it removes the states. Can you attach the codes that describe the situation you describe?

Thank you for sharing your opinion :)

alexjeffburke commented:
Will attach the code when I’m back at my desk :)

But, I think you slightly misunderstood my comments about FAILURE. Receiving a result from celery isn’t enough to know it was successful, you have to check the response message `.status === 'SUCCESS'`

Put more simply, if the AsyncResult rejects on a TIMEOUT then as a user I’d probably also expect it to reject on a FAILURE message from celery :)

actumn commented:
Ah, I see. I didn't catch that celery throws an exception when it fails. 

> Put more simply, if the AsyncResult rejects on a TIMEOUT then as a user I’d probably also expect it to reject on a FAILURE message from celery :)

I think that's perfect  :+1:. Yeah, it should reject on a FAILURE message.

alexjeffburke commented on src/app/result.ts:
```diff
@@ -15,25 +22,26 @@ export class AsyncResult {
-      if (this.result) {
-        resolve(this.result);
-      }
+    if (this._promise) {
```
I've gone into more detail about this change in the comment message for https://github.com/actumn/celery.node/pull/25/commits/d69de37733f567b5615056b9414aa06afd835387

alexjeffburke commented:
@actumn thanks for the feedback - I've made the conversion to reject the promise on timeout/failure message and removed the status completely. In doing so I found there was a small issue with multiple calls to `.get()` so I've addressed that and covered it with a test too.

actumn reviewed (approved):
Looks good to me!

alexjeffburke commented:
@actumn thanks for the merge in and release :)
